src/common/components/markdownViewer/syncScroll/synchronizeScroll.ts

```diff
@@ -0,0 +1,113 @@
+
+const SOURCE_LINE_CLASSNAME = 'sourceLine';
+const SOURCE_LINE_ATTRIBUTE = 'line';
+const SOURCE_LINE_HIGHLIGHT_CLASSNAME = 'highlightSourceLine';
+
+interface Candidate {
+  lineNum: number;
+  element: Element;
+}
+
+interface CandidateSearch {
+  exactMatch: boolean;
+  candidate: Candidate;
+  nextCandidate?: Candidate;
+}
+
+const isIntegerProperty = (property: string) => {
+  return Number.isInteger(Number(property));
+};
+
+// Retrieve the candidate(s) elements that correspond to a certain
+// source line for an input list of rendered elements.
+const getRenderedElementsForSourceLine = (elementList: NodeListOf<Element>, targetLine: number): CandidateSearch => {
+  let candidate: Candidate = null;
+  if (elementList[0] && isIntegerProperty(elementList[0].getAttribute(SOURCE_LINE_ATTRIBUTE))) {
+    candidate = {
+      lineNum: Number(elementList[0].getAttribute(SOURCE_LINE_ATTRIBUTE)),
+      element: elementList[0],
+    };
+  }
+
+  for (const element of elementList) {
+    const slAttr = element.getAttribute(SOURCE_LINE_ATTRIBUTE);
+    if (isIntegerProperty(slAttr)) {
+      const current: Candidate = { lineNum: Number(slAttr), element };
+      if (current.lineNum === targetLine) {
+        return { exactMatch: true, candidate: current };
+      }
+      else if (current.lineNum > targetLine) {
+        return { exactMatch: false, candidate, nextCandidate: current };
+      }
+      candidate = current;
+    }
+  }
+  return { exactMatch: false, candidate };
+};
+
+// Retrieve the candidate(s) elements that are located in a certain pixel offset
+// position from an input list of rendered elements.
+const getRenderedElementsForPixelOffset = (elementList: NodeListOf<Element>, targetOffset: number): CandidateSearch => {
+    let candidate: Candidate = null;
+    let lineFraction = 0;
+    for (const element of elementList) {
+      const slAttr = element.getAttribute(SOURCE_LINE_ATTRIBUTE);
+      if (isIntegerProperty(slAttr)) {
+        const boundingRect = element.getBoundingClientRect();
+        const current: Candidate = { lineNum: Number(slAttr), element };
+        if (boundingRect.top > targetOffset) {
+          if (candidate && lineFraction < 1) {
+            candidate.lineNum += lineFraction;
+            return { exactMatch: true, candidate };
+          }
+          return { exactMatch: false, candidate, nextCandidate: current };
+        }
+        lineFraction = (targetOffset - boundingRect.top) / boundingRect.height;
+        candidate = current;
+      }
+    }
+    return { exactMatch: false, candidate };
+};
```

Comment:
[Kameame / Adouken](http://i.imgur.com/BtjZedW.jpg) or more formar cyclomatic complexity.

This should be read like a book following the same level of abstraction principle.

If you can tell us briefly what this method does we will make a refactor proposal (pseudocode)

Reply:
Same here, please provide pseucode what this method does and we will come up with a clean code proposal.

Reply:
Check refactor as commented above. [Link to commit](https://github.com/Lemoncode/LeanMood/pull/168/commits/d9cd50cd321fb25e7f3fa0159abe6bb07c3ef661).